File: Sources/OrgRecCore/VAO04Runtime.swift
import Foundation

public enum VAO04RuntimeError: Error, LocalizedError, Equatable {
    case missingRandomSource(String)
    case unsupportedRandomAlgorithm(String)
    case invalidDistribution(String)
    case conflictingWrite(String)
    case maximumMicrosteps

    public var errorDescription: String? {
        switch self {
        case .missingRandomSource(let id): "Missing reproducible random source \(id)."
        case .unsupportedRandomAlgorithm(let value): "Unsupported random algorithm \(value)."
        case .invalidDistribution(let id): "Invalid probability distribution for \(id)."
        case .conflictingWrite(let id): "Rejected conflicting write to \(id)."
        case .maximumMicrosteps: "The deterministic runtime exceeded maximumMicrosteps."
        }
    }
}

public struct VAO04RuntimeResult: Codable, Sendable, Equatable {
    public var state: [String: VAOJSONValue]
    public var emittedEvents: [VAOJSONValue]
    public var renderBindingIds: [String]
}

private struct VAO04PCG32: Sendable {
    var state: UInt64
    let increment: UInt64

    init(seedHex: String, stream: Int) {
        state = 0
        increment = (UInt64(stream) << 1) | 1
        _ = next()
        state &+= UInt64(seedHex.suffix(16), radix: 16) ?? 0
        _ = next()
    }

    mutating func next() -> UInt32 {
        let old = state
        state = old &* 6_364_136_223_846_793_005 &+ increment
        let shifted = UInt32(truncatingIfNeeded: ((old >> 18) ^ old) >> 27)
        let rotation = UInt32(old >> 59)
        return (shifted >> rotation) | (shifted << ((0 &- rotation) & 31))
    }

    mutating func uniform() -> Double { Double(next()) / 4_294_967_296.0 }
}

public enum VAO04RuntimeEngine {
    private struct OrderedAction {
        var group: String; var index: Int; var action: VAO03DeclarativeAction; var transition: VAO03InteractionTransition
    }

    private static func condition(_ condition: VAO03StateCondition, state: [String: VAOJSONValue]) -> Bool {
        let actual = state[condition.stateVariableId]
        return switch condition.comparisonOperator {
        case "equals": actual == condition.value
        case "not-equals": actual != condition.value
        case "less-than": (actual?.numberValue ?? .infinity) < (condition.value.numberValue ?? -.infinity)
        case "less-or-equal": (actual?.numberValue ?? .infinity) <= (condition.value.numberValue ?? -.infinity)
        case "greater-than": (actual?.numberValue ?? -.infinity) > (condition.value.numberValue ?? .infinity)
        case "greater-or-equal": (actual?.numberValue ?? -.infinity) >= (condition.value.numberValue ?? .infinity)
        default: false
        }
    }

    public static func execute(manifest: VAO04Manifest, events: [VAO04TraceEvent], initialState: [String: VAOJSONValue] = [:]) throws -> VAO04RuntimeResult {
        guard let model = manifest.interactionModel else {
            return VAO04RuntimeResult(state: initialState, emittedEvents: [], renderBindingIds: [])
        }
        var state = Dictionary(uniqueKeysWithValues: model.stateVariables.map { ($0.id, $0.defaultValue) })
        state.merge(initialState) { _, supplied in supplied }
        let transitions = model.transitions.sorted { lhs, rhs in
            let lp = lhs.priority ?? 0, rp = rhs.priority ?? 0
            return lp == rp ? lhs.id < rhs.id : lp > rp
        }
        let processes = Dictionary(uniqueKeysWithValues: model.processModels.map { ($0.id, $0) })
        var random: [String: VAO04PCG32] = [:]
        for record in manifest.runtime.randomSources + (model.randomSources ?? []) {
            guard record.algorithm == "pcg32" else { throw VAO04RuntimeError.unsupportedRandomAlgorithm(record.algorithm) }
            random[record.id] = VAO04PCG32(seedHex: record.seed, stream: record.stream)
        }
        let orderedEvents = events.sorted { lhs, rhs in
            if lhs.timestamp != rhs.timestamp { return lhs.timestamp < rhs.timestamp }
            if (lhs.priority ?? 0) != (rhs.priority ?? 0) { return (lhs.priority ?? 0) > (rhs.priority ?? 0) }
            if lhs.eventTypeId != rhs.eventTypeId { return lhs.eventTypeId < rhs.eventTypeId }
            return lhs.sequence < rhs.sequence
        }
        var emitted: [VAOJSONValue] = []
        var selected: [String] = []
        var microsteps = 0
        let maximum = model.executionSemantics.maximumMicrosteps ?? 10_000

        func processActions(_ process: VAO04ProcessModel) throws -> [VAO03DeclarativeAction] {
            var actions = process.actions
            for child in process.childProcessIds ?? [] {
                if let record = processes[child] { actions += try processActions(record) }
            }
            guard process.processKind == "stochastic" else { return actions }
            guard let randomID = process.randomSourceId, var generator = random[randomID] else { throw VAO04RuntimeError.missingRandomSource(process.randomSourceId ?? "") }
            defer { random[randomID] = generator }
            guard !actions.isEmpty, let distribution = process.probabilityDistribution else { throw VAO04RuntimeError.invalidDistribution(process.id) }
            if distribution.kind == "categorical" {
                let weights = actions.indices.map { distribution.parameters[String($0)] ?? 0 }
                let total = weights.reduce(0, +)
                guard total > 0 else { throw VAO04RuntimeError.invalidDistribution(process.id) }
                let threshold = generator.uniform() * total
                var running = 0.0
                for (action, weight) in zip(actions, weights) { running += weight; if threshold < running { return [action] } }
                return [actions.last!]
            }
            return [actions[min(Int(generator.uniform() * Double(actions.count)), actions.count - 1)]]
        }

        for event in orderedEvents {
            let snapshot = state
            let matching = transitions.filter { transition in
                transition.eventTypeId == event.eventTypeId
                    && (transition.controlId == nil || transition.controlId == event.controlId)
                    && (transition.conditions ?? []).allSatisfy { condition($0, state: snapshot) }
            }
            var actions: [OrderedAction] = []
            for transition in matching {
                for (index, action) in transition.actions.enumerated() {
                    actions.append(OrderedAction(group: action.executionGroup ?? "", index: index, action: action, transition: transition))
                }
            }
            actions.sort { lhs, rhs in
                if lhs.group != rhs.group { return lhs.group < rhs.group }
                if lhs.index != rhs.index { return lhs.index < rhs.index }
                return lhs.transition.id < rhs.transition.id
            }
            var writes: [String: VAOJSONValue] = [:]
            for ordered in actions {
                microsteps += 1; if microsteps > maximum { throw VAO04RuntimeError.maximumMicrosteps }
                let action = ordered.action, target = action.targetId
                switch action.operation {
                case "set-state", "toggle-state", "increment-state":
                    let current = writes[target] ?? snapshot[target]
                    let next: VAOJSONValue
                    if action.operation == "toggle-state" { next = .boolean(!(current?.booleanValue ?? false)) }
                    else if action.operation == "increment-state" { next = .number((current?.numberValue ?? 0) + (action.value?.numberValue ?? 0)) }
                    else { next = action.value ?? .null }
                    if let prior = writes[target], prior != next, ordered.transition.conflictPolicy == "reject" { throw VAO04RuntimeError.conflictingWrite(target) }
                    writes[target] = next
                case "emit-event": emitted.append(.object(["eventTypeId": .string(target), "timestamp": .number(event.timestamp), "sourceTransitionId": .string(ordered.transition.id)]))
                case "route-event": emitted.append(.object(["routeId": .string(target), "timestamp": .number(event.timestamp), "sourceTransitionId": .string(ordered.transition.id)]))
                case "start-process":
                    if let process = processes[target] {
                        for processAction in try processActions(process) {
                            emitted.append(.object(["processId": .string(target), "operation": .string(processAction.operation), "targetId": .string(processAction.targetId), "timestamp": .number(event.timestamp)]))
                        }
                    }
                case "select-render-binding": selected.append(target)
                default: break
                }
            }
            state.merge(writes) { _, new in new }
            for binding in model.renderBindings where binding.eventTypeId == nil || binding.eventTypeId == event.eventTypeId {
                if (binding.conditions ?? []).allSatisfy({ condition($0, state: state) }) { selected.append(binding.id) }
            }
        }
        var seen: Set<String> = []
        let uniqueSelected = selected.filter { seen.insert($0).inserted }
        return VAO04RuntimeResult(state: state, emittedEvents: emitted, renderBindingIds: uniqueSelected)
    }

    public static func verify(trace: VAO04ConformanceTrace, manifest: VAO04Manifest) -> [String] {
        var errors: [String] = []
        do {
            let result = try execute(manifest: manifest, events: trace.inputEvents, initialState: trace.initialState ?? [:])
            let expected = VAO04RuntimeResult(state: trace.expected.state, emittedEvents: trace.expected.emittedEvents, renderBindingIds: trace.expected.renderBindingIds)
            if result != expected { errors.append("Conformance trace \(trace.id) does not match deterministic execution.") }
        } catch { errors.append("Conformance trace \(trace.id) cannot execute: \(error.localizedDescription)") }
        return errors
    }
}

File: Tools/vao04_runtime.py
#!/usr/bin/env python3
"""Deterministic, declarative VAO 0.4.0 interaction reference interpreter."""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from typing import Any


def canonical_bytes(value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"), allow_nan=False).encode("utf-8")


@dataclass
class PCG32:
    state: int
    increment: int

    @classmethod
    def from_record(cls, record: dict[str, Any]) -> "PCG32":
        seed = int(record["seed"], 16) & ((1 << 64) - 1)
        stream = int(record["stream"])
        generator = cls(0, ((stream << 1) | 1) & ((1 << 64) - 1))
        generator.next_uint32()
        generator.state = (generator.state + seed) & ((1 << 64) - 1)
        generator.next_uint32()
        return generator

    def next_uint32(self) -> int:
        old = self.state
        self.state = (old * 6364136223846793005 + self.increment) & ((1 << 64) - 1)
        xorshifted = (((old >> 18) ^ old) >> 27) & 0xFFFFFFFF
        rotation = (old >> 59) & 31
        return ((xorshifted >> rotation) | (xorshifted << ((-rotation) & 31))) & 0xFFFFFFFF

    def uniform(self) -> float:
        return self.next_uint32() / 4294967296.0


def condition_matches(condition: dict[str, Any], state: dict[str, Any]) -> bool:
    actual = state.get(condition["stateVariableId"])
    expected = condition.get("value")
    operation = condition["operator"]
    if operation == "equals": return actual == expected
    if operation == "not-equals": return actual != expected
    if operation == "less-than": return actual is not None and actual < expected
    if operation == "less-or-equal": return actual is not None and actual <= expected
    if operation == "greater-than": return actual is not None and actual > expected
    if operation == "greater-or-equal": return actual is not None and actual >= expected
    if operation == "contains": return isinstance(actual, (list, str)) and expected in actual
    return False


class RuntimeError04(ValueError):
    pass


class Interpreter:
    def __init__(self, manifest: dict[str, Any]):
        self.manifest = manifest
        self.model = manifest.get("interactionModel") or {}
        self.runtime = manifest["runtime"]
        self.semantics = self.runtime["executionSemantics"]
        self.events = {x["id"]: x for x in self.model.get("eventTypes", [])}
        self.transitions = sorted(self.model.get("transitions", []), key=lambda x: (-x.get("priority", 0), x["id"]))
        self.processes = {x["id"]: x for x in self.model.get("processModels", [])}
        self.routes = {x["id"]: x for x in self.model.get("routingRules", [])}
        self.renders = self.model.get("renderBindings", [])
        random_records = {x["id"]: x for x in self.runtime.get("randomSources", [])}
        random_records.update({x["id"]: x for x in self.model.get("randomSources", [])})
        self.random = {key: PCG32.from_record(value) for key, value in random_records.items() if value["algorithm"] == "pcg32"}
        self.maximum_microsteps = self.semantics.get("maximumMicrosteps", 10_000)

    def initial_state(self, override: dict[str, Any] | None = None) -> dict[str, Any]:
        state = {record["id"]: record["defaultValue"] for record in self.model.get("stateVariables", [])}
        state.update(override or {})
        return state

    def _process_actions(self, process: dict[str, Any]) -> list[dict[str, Any]]:
        actions = list(process.get("actions", []))
        for child in process.get("childProcessIds", []):
            actions.extend(self._process_actions(self.processes[child]))
        if process.get("processKind") != "stochastic":
            return actions
        generator = self.random.get(process.get("randomSourceId"))
        if generator is None:
            raise RuntimeError04(f"Missing reproducible random source for {process['id']}")
        distribution = process["probabilityDistribution"]
        if distribution["kind"] == "categorical":
            weights = [distribution["parameters"].get(str(index), 0.0) for index in range(len(actions))]
            total = sum(weights)
            if total <= 0:
                raise RuntimeError04(f"Categorical process {process['id']} has no positive weight")
            threshold = generator.uniform() * total
            running = 0.0
            for action, weight in zip(actions, weights):
                running += weight
                if threshold < running:
                    return [action]
            return [actions[-1]]
        index = min(int(generator.uniform() * len(actions)), len(actions) - 1)
        return [actions[index]]

    def execute(self, events: list[dict[str, Any]], initial_state: dict[str, Any] | None = None) -> dict[str, Any]:
        state = self.initial_state(initial_state)
        emitted: list[dict[str, Any]] = []
        selected: list[str] = []
        ordered = sorted(events, key=lambda event: (
            event["timestamp"], -event.get("priority", 0), event["eventTypeId"], event["sequence"]
        ))
        microsteps = 0
        for event in ordered:
            snapshot = dict(state)
            matching = [transition for transition in self.transitions
                        if transition["eventTypeId"] == event["eventTypeId"]
                        and (transition.get("controlId") is None or transition.get("controlId") == event.get("controlId"))
                        and all(condition_matches(c, snapshot) for c in transition.get("conditions", []))]
            writes: dict[str, Any] = {}
            actions: list[tuple[str, int, dict[str, Any], dict[str, Any]]] = []
            for transition in matching:
                for index, action in enumerate(transition["actions"]):
                    actions.append((action.get("executionGroup", ""), index, action, transition))
            for _, _, action, transition in sorted(actions, key=lambda item: (item[0], item[1], item[3]["id"])):
                microsteps += 1
                if microsteps > self.maximum_microsteps:
                    raise RuntimeError04("maximumMicrosteps exceeded")
                operation, target = action["operation"], action["targetId"]
                if operation in {"set-state", "toggle-state", "increment-state"}:
                    current = writes.get(target, snapshot.get(target))
                    value = action.get("value")
                    if operation == "toggle-state": value = not bool(current)
                    if operation == "increment-state": value = current + value
                    if target in writes and writes[target] != value and transition.get("conflictPolicy") == "reject":
                        raise RuntimeError04(f"Rejected conflicting write to {target}")
                    writes[target] = value
                elif operation == "emit-event":
                    emitted.append({"eventTypeId": target, "timestamp": event["timestamp"], "sourceTransitionId": transition["id"]})
                elif operation == "route-event":
                    emitted.append({"routeId": target, "timestamp": event["timestamp"], "sourceTransitionId": transition["id"]})
                elif operation == "start-process":
                    for process_action in self._process_actions(self.processes[target]):
                        emitted.append({"processId": target, "operation": process_action["operation"], "targetId": process_action["targetId"], "timestamp": event["timestamp"]})
                elif operation == "select-render-binding":
                    selected.append(target)
            state.update(writes)
            for binding in self.renders:
                if binding.get("eventTypeId") not in (None, event["eventTypeId"]): continue
                if all(condition_matches(c, state) for c in binding.get("conditions", [])):
                    selected.append(binding["id"])
        result = {
            "state": state,
            "emittedEvents": emitted,
            "renderBindingIds": list(dict.fromkeys(selected)),
        }
        return result


def trace_digest(initial_state: dict[str, Any], input_events: list[dict[str, Any]], expected: dict[str, Any]) -> str:
    return hashlib.sha256(canonical_bytes({"initialState": initial_state, "inputEvents": input_events, "expected": expected})).hexdigest()


def verify_trace(manifest: dict[str, Any], trace: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    expected_digest = trace_digest(trace.get("initialState", {}), trace["inputEvents"], trace["expected"])
    digest = trace["digest"]
    if digest["algorithm"] != "sha256" or digest["value"] != expected_digest:
        errors.append(f"Conformance trace {trace['id']} has a non-canonical digest.")
    try:
        actual = Interpreter(manifest).execute(trace["inputEvents"], trace.get("initialState"))
    except (KeyError, TypeError, RuntimeError04) as exc:
        errors.append(f"Conformance trace {trace['id']} cannot execute: {exc}")
        return errors
    if actual != trace["expected"]:
        errors.append(f"Conformance trace {trace['id']} does not match deterministic execution.")
    return errors
